File: sum.c
#include <dlfcn.h>
#include <stdio.h>
#include <stdlib.h>

#define LLVMModuleRef void*
#define LLVMTypeRef void*
#define LLVMValueRef void*
#define LLVMBasicBlockRef void*
#define LLVMBuilderRef void*
#define LLVMExecutionEngineRef void*

#define LLVMAbortProcessAction 0

void load_llvm();

void* (*LLVMModuleCreateWithName)(const char*);
void* (*LLVMInt32Type)();
void* (*LLVMFunctionType)(LLVMTypeRef, LLVMTypeRef*, unsigned, unsigned);
void* (*LLVMAddFunction)(LLVMModuleRef, const char*, LLVMTypeRef);
void* (*LLVMAppendBasicBlock)(LLVMValueRef, const char*);
void* (*LLVMCreateBuilder)();
void (*LLVMPositionBuilderAtEnd)(LLVMBuilderRef, LLVMBasicBlockRef);
void* (*LLVMBuildAdd)(LLVMBuilderRef, LLVMValueRef, LLVMValueRef, const char*);
void* (*LLVMGetParam)(LLVMValueRef, unsigned);
void* (*LLVMBuildRet)(LLVMBuilderRef, LLVMValueRef);
void* (*LLVMVerifyModule)(LLVMModuleRef, unsigned, char**);
void (*LLVMDisposeMessage)(char*);
void (*LLVMDumpModule)(LLVMModuleRef);
void (*LLVMLinkInMCJIT)();
int (*LLVMInitializeX86Target)();
int (*LLVMInitializeX86AsmPrinter)();
int (*LLVMCreateExecutionEngineForModule)(LLVMExecutionEngineRef*, LLVMModuleRef, char**);
void* (*LLVMGetFunctionAddress)(LLVMExecutionEngineRef, const char*);
void (*LLVMDisposeBuilder)(LLVMBuilderRef);
void (*LLVMDisposeExecutionEngine)(LLVMExecutionEngineRef);

int main() {
    load_llvm();

    LLVMModuleRef mod = LLVMModuleCreateWithName("my_module");

    LLVMTypeRef param_types[] = { LLVMInt32Type(), LLVMInt32Type() };
    LLVMTypeRef ret_type = LLVMFunctionType(LLVMInt32Type(), param_types, 2, 0);
    LLVMValueRef sum = LLVMAddFunction(mod, "sum", ret_type);

    LLVMBasicBlockRef entry = LLVMAppendBasicBlock(sum, "entry");

    LLVMBuilderRef builder = LLVMCreateBuilder();
    LLVMPositionBuilderAtEnd(builder, entry);
    LLVMValueRef tmp = LLVMBuildAdd(builder, LLVMGetParam(sum, 0), LLVMGetParam(sum, 1), "tmp");
    LLVMBuildRet(builder, tmp);

    char *error = NULL;
    LLVMVerifyModule(mod, LLVMAbortProcessAction, &error);
    LLVMDisposeMessage(error);

    LLVMDumpModule(mod);

    LLVMExecutionEngineRef engine;
    error = NULL;
    LLVMLinkInMCJIT();
    LLVMInitializeX86Target();
    LLVMInitializeX86AsmPrinter();
    if (LLVMCreateExecutionEngineForModule(&engine, mod, &error) != 0) {
        fprintf(stderr, "failed to create execution engine\n");
        abort();
    }
    if (error) {
        fprintf(stderr, "error: %s\n", error);
        LLVMDisposeMessage(error);
        exit(EXIT_FAILURE);
    }

    int (*sum_func)(int, int) = (int (*)(int, int))LLVMGetFunctionAddress(engine, "sum");
    printf("%d\n", sum_func(34, 35));

    LLVMDisposeBuilder(builder);
    LLVMDisposeExecutionEngine(engine);
}

#define ADD_DLSYM(name) (name = dlsym(handle, #name))

void load_llvm() {
    void *handle = dlopen("./libLLVM-15git.so", RTLD_LAZY | RTLD_DEEPBIND);
    if (!handle) {
        fprintf(stderr, "failed to load libLLVM\n");
        abort();
    }

    ADD_DLSYM(LLVMModuleCreateWithName);
    ADD_DLSYM(LLVMInt32Type);
    ADD_DLSYM(LLVMFunctionType);
    ADD_DLSYM(LLVMAddFunction);
    ADD_DLSYM(LLVMAppendBasicBlock);
    ADD_DLSYM(LLVMCreateBuilder);
    ADD_DLSYM(LLVMPositionBuilderAtEnd);
    ADD_DLSYM(LLVMBuildAdd);
    ADD_DLSYM(LLVMGetParam);
    ADD_DLSYM(LLVMBuildRet);
    ADD_DLSYM(LLVMVerifyModule);
    ADD_DLSYM(LLVMDisposeMessage);
    ADD_DLSYM(LLVMDumpModule);
    ADD_DLSYM(LLVMLinkInMCJIT);
    ADD_DLSYM(LLVMInitializeX86Target);
    ADD_DLSYM(LLVMInitializeX86AsmPrinter);
    ADD_DLSYM(LLVMCreateExecutionEngineForModule);
    ADD_DLSYM(LLVMGetFunctionAddress);
    ADD_DLSYM(LLVMDisposeBuilder);
    ADD_DLSYM(LLVMDisposeExecutionEngine);
}

File: sum.ts
const ptr = "pointer";
const LLVMAbortProcessAction = 0;

const llvm = loadLLVM();

const mod = llvm.LLVMModuleCreateWithName(cstr("my_module"));

const param_types = [llvm.LLVMInt32Type(), llvm.LLVMInt32Type()];
const ret_type = llvm.LLVMFunctionType(
  llvm.LLVMInt32Type(),
  ptrs(param_types),
  2,
  0,
);
const sum = llvm.LLVMAddFunction(mod, cstr("sum"), ret_type);

const entry = llvm.LLVMAppendBasicBlock(sum, cstr("entry"));

const builder = llvm.LLVMCreateBuilder();
llvm.LLVMPositionBuilderAtEnd(builder, entry);
const tmp = llvm.LLVMBuildAdd(
  builder,
  llvm.LLVMGetParam(sum, 0),
  llvm.LLVMGetParam(sum, 1),
  cstr("tmp"),
);
llvm.LLVMBuildRet(builder, tmp);

let error = new BigUint64Array(1);
llvm.LLVMVerifyModule(mod, LLVMAbortProcessAction, error);
llvm.LLVMDisposeMessage(error[0]);

llvm.LLVMDumpModule(mod);

const engine = new BigUint64Array(1);
error = new BigUint64Array(1);

llvm.LLVMLinkInMCJIT();

// LLVMInitializeNativeTarget(); is a macro for following 3 calls:
llvm.LLVMInitializeX86TargetInfo();
llvm.LLVMInitializeX86Target();
llvm.LLVMInitializeX86TargetMC();

// LLVMInitializeNativeAsmPrinter(); is a macro for following call:
llvm.LLVMInitializeX86AsmPrinter();

if (llvm.LLVMCreateExecutionEngineForModule(engine, mod, error) != 0) {
  console.error("failed to create execution engine");
  Deno.exit();
}
if (error[0]) {
  console.error("error:", fromCString(error[0]));
  llvm.LLVMDisposeMessage(error[0]);
  Deno.exit(1);
}

const sum_func = new Deno.UnsafeFnPointer(
  llvm.LLVMGetFunctionAddress(engine[0], cstr("sum")),
  {
    parameters: ["i32", "i32"],
    result: "i32",
  },
);
console.log(sum_func.call(34, 35));

llvm.LLVMDisposeBuilder(builder);
llvm.LLVMDisposeExecutionEngine(engine[0]);

// utils

function cstr(str: string) {
  return Uint8Array.from(toCharCodes(str + "\0"));
}

function toCharCodes(str: string): number[] {
  return [...str].map((_, i) => str.charCodeAt(i));
}

function fromCString(pointer: bigint) {
  return new Deno.UnsafePointerView(pointer).getCString();
}

function ptrs(pointers: bigint[]) {
  return Deno.UnsafePointer.of(
    BigInt64Array.from(pointers),
  );
}

function loadLLVM() {
  const lib = Deno.dlopen("./libLLVM-15git.so", {
    LLVMModuleCreateWithName: {
      parameters: [ptr],
      result: ptr,
    },
    LLVMInt32Type: {
      parameters: [],
      result: ptr,
    },
    LLVMFunctionType: {
      parameters: [ptr, ptr, "i32", "i32"],
      result: ptr,
    },
    LLVMAddFunction: {
      parameters: [ptr, ptr, ptr],
      result: ptr,
    },
    LLVMAppendBasicBlock: {
      parameters: [ptr, ptr],
      result: ptr,
    },
    LLVMCreateBuilder: {
      parameters: [],
      result: ptr,
    },
    LLVMPositionBuilderAtEnd: {
      parameters: [ptr, ptr],
      result: ptr,
    },
    LLVMBuildAdd: {
      parameters: [ptr, ptr, ptr, ptr],
      result: ptr,
    },
    LLVMGetParam: {
      parameters: [ptr, "i32"],
      result: ptr,
    },
    LLVMBuildRet: {
      parameters: [ptr, ptr],
      result: ptr,
    },
    LLVMVerifyModule: {
      parameters: [ptr, "i32", ptr],
      result: ptr,
    },
    LLVMDisposeMessage: {
      parameters: [ptr],
      result: "void",
    },
    LLVMDumpModule: {
      parameters: [ptr],
      result: "void",
    },
    LLVMLinkInMCJIT: {
      parameters: [],
      result: "void",
    },
    LLVMInitializeX86TargetInfo: {
      parameters: [],
      result: "i32",
    },
    LLVMInitializeX86Target: {
      parameters: [],
      result: "i32",
    },
    LLVMInitializeX86TargetMC: {
      parameters: [],
      result: "i32",
    },
    LLVMInitializeX86AsmPrinter: {
      parameters: [],
      result: "i32",
    },
    LLVMCreateExecutionEngineForModule: {
      parameters: [ptr, ptr, ptr],
      result: "i32",
    },
    LLVMGetFunctionAddress: {
      parameters: [ptr, ptr],
      result: "function",
    },
    LLVMDisposeBuilder: {
      parameters: [ptr],
      result: "void",
    },
    LLVMDisposeExecutionEngine: {
      parameters: [ptr],
      result: "void",
    },
  });

  return lib.symbols;
}
